From e61293d2325dbe3d3435619bda985a01b0557cc7 Mon Sep 17 00:00:00 2001
From: Fugang Duan <b38611@freescale.com>
Date: Wed, 5 Nov 2014 13:13:02 +0800
Subject: [PATCH] MLK-9798 tty: serial: imx: fix the dma overwrite and buffer
 index issue

This reverts commit 0f7c43a163521ea081d7743c9a55314e7deba728.
And add another change that mod the variable "last_completed_idx"
after it increasing.

Signed-off-by: Fugang Duan <B38611@freescale.com>
---
 drivers/tty/serial/imx.c | 27 +++++++++++++++------------
 1 file changed, 15 insertions(+), 12 deletions(-)

diff --git a/drivers/tty/serial/imx.c b/drivers/tty/serial/imx.c
index 61fc09c..2f2a413 100644
--- a/drivers/tty/serial/imx.c
+++ b/drivers/tty/serial/imx.c
@@ -921,13 +921,13 @@ static void dma_rx_work(struct work_struct *w)
 
 	if (sport->rx_buf.last_completed_idx < sport->rx_buf.cur_idx) {
 		dma_rx_push_data(sport, tty, sport->rx_buf.last_completed_idx + 1,
-					sport->rx_buf.cur_idx + 1);
+					sport->rx_buf.cur_idx);
 	} else if (sport->rx_buf.last_completed_idx == (IMX_RXBD_NUM - 1)) {
-		dma_rx_push_data(sport, tty, 0, sport->rx_buf.cur_idx + 1);
+		dma_rx_push_data(sport, tty, 0, sport->rx_buf.cur_idx);
 	} else {
 		dma_rx_push_data(sport, tty, sport->rx_buf.last_completed_idx + 1,
 					IMX_RXBD_NUM);
-		dma_rx_push_data(sport, tty, 0, sport->rx_buf.cur_idx + 1);
+		dma_rx_push_data(sport, tty, 0, sport->rx_buf.cur_idx);
 	}
 }
 
@@ -975,17 +975,20 @@ static void dma_rx_callback(void *data)
 		writel(USR2_IDLE, sport->port.membase + USR2);
 	}
 
-	if (count > 0) {
-		sport->rx_buf.buf_info[sport->rx_buf.cur_idx].filled = true;
-		sport->rx_buf.buf_info[sport->rx_buf.cur_idx].rx_bytes = count;
-		sport->rx_buf.cur_idx++;
-		sport->rx_buf.cur_idx %= IMX_RXBD_NUM;
-		dev_dbg(sport->port.dev, "We get %d bytes.\n", count);
+	sport->rx_buf.buf_info[sport->rx_buf.cur_idx].filled = true;
+	sport->rx_buf.buf_info[sport->rx_buf.cur_idx].rx_bytes = count;
+	sport->rx_buf.cur_idx++;
+	sport->rx_buf.cur_idx %= IMX_RXBD_NUM;
+	dev_dbg(sport->port.dev, "We get %d bytes.\n", count);
 
-		if (sport->rx_buf.cur_idx == sport->rx_buf.last_completed_idx)
-			dev_err(sport->port.dev, "overwrite!\n");
+	if (sport->rx_buf.cur_idx == sport->rx_buf.last_completed_idx)
+		dev_err(sport->port.dev, "overwrite!\n");
 
+	if (count) {
 		schedule_work(&sport->tsk_dma_rx);
+	} else {
+		sport->rx_buf.last_completed_idx++;
+		sport->rx_buf.last_completed_idx %= IMX_RXBD_NUM;
 	}
 }
 
@@ -998,7 +1001,7 @@ static int start_rx_dma(struct imx_port *sport)
 	sport->rx_buf.period_len = RX_BUF_SIZE;
 	sport->rx_buf.buf_len = IMX_RXBD_NUM * RX_BUF_SIZE;
 	sport->rx_buf.cur_idx = 0;
-	sport->rx_buf.last_completed_idx = 0;
+	sport->rx_buf.last_completed_idx = -1;
 	desc = dmaengine_prep_dma_cyclic(chan, sport->rx_buf.dmaaddr,
 		sport->rx_buf.buf_len, sport->rx_buf.period_len,
 		DMA_DEV_TO_MEM, DMA_PREP_INTERRUPT);
-- 
1.8.0

